File: user-service/user-api/src/main/java/com/yellowsunn/userservice/filter/LoginUserArgumentResolver.java
package com.yellowsunn.userservice.annotation;

import com.yellowsunn.common.annotation.LoginUser;
import com.yellowsunn.common.exception.InvalidAuthenticationException;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.core.MethodParameter;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

import static com.yellowsunn.common.constant.CommonHeaderConst.USER_UUID_HEADER;

@Slf4j
public class LoginUserArgumentResolver implements HandlerMethodArgumentResolver {
    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.hasParameterAnnotation(LoginUser.class);
    }

    @Override
    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
        String uuid = webRequest.getHeader(USER_UUID_HEADER);
        LoginUser loginUser = parameter.getParameterAnnotation(LoginUser.class);
        if (loginUser != null && loginUser.required() && StringUtils.isBlank(uuid)) {
            throw new InvalidAuthenticationException();
        }
        return uuid;
    }
}

File: board-service/apps/board-api/src/main/kotlin/com/yellowsunn/boardservice/filter/LoginUserArgumentResolver.kt
package com.yellowsunn.boardservice.filter

import com.yellowsunn.common.annotation.LoginUser
import com.yellowsunn.common.constant.CommonHeaderConst.USER_UUID_HEADER
import com.yellowsunn.common.exception.InvalidAuthenticationException
import org.springframework.core.MethodParameter
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer

class LoginUserArgumentResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return parameter.hasParameterAnnotation(LoginUser::class.java)
    }

    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?,
    ): Any? {
        val uuid: String? = webRequest.getHeader(USER_UUID_HEADER)
        val loginUser: LoginUser? = parameter.getParameterAnnotation(LoginUser::class.java)
        if (loginUser?.required == true && uuid.isNullOrBlank()) {
            throw InvalidAuthenticationException()
        }
        return uuid
    }
}
